sql/core/src/main/scala/org/apache/spark/sql/execution/analysis/ResolveDatasetColumnReference.scala

```diff
@@ -0,0 +1,136 @@
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, SubqueryAlias}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.MetadataBuilder
+
+/**
+ * Resolves the Dataset column reference by traversing the query plan and finding the plan subtree
+ * of the Dataset that the column reference belongs to.
+ *
+ * Dataset column reference is simply an [[AttributeReference]] that is returned by `Dataset#col`.
+ * Most of time we don't need to do anything special, as [[AttributeReference]] can point to
+ * the column precisely. However, in case of self-join, the analyzer generates
+ * [[AttributeReference]] with new expr IDs for the right side plan of the join. If the Dataset
+ * column reference points to a column in the right side plan of a self-join, we need to replace it
+ * with the corresponding newly generated [[AttributeReference]].
+ */
+class ResolveDatasetColumnReference(conf: SQLConf) extends Rule[LogicalPlan] {
+
+  // Dataset column reference is an `AttributeReference` with 2 special metadata.
+  private def isColumnReference(a: AttributeReference): Boolean = {
+    a.metadata.contains(Dataset.ID_PREFIX) && a.metadata.contains(Dataset.COL_POS_PREFIX)
+  }
+
+  private case class ColumnReference(datasetId: Long, colPos: Int)
+
+  private def toColumnReference(a: AttributeReference): ColumnReference = {
+    ColumnReference(
+      a.metadata.getLong(Dataset.ID_PREFIX),
+      a.metadata.getLong(Dataset.COL_POS_PREFIX).toInt)
+  }
+
+  private def stripColumnReferenceMetadata(a: AttributeReference): AttributeReference = {
+    val metadataWithoutId = new MetadataBuilder()
+      .withMetadata(a.metadata)
+      .remove(Dataset.ID_PREFIX)
+      .remove(Dataset.COL_POS_PREFIX)
+      .build()
+    a.withMetadata(metadataWithoutId)
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!conf.getConf(SQLConf.RESOLVE_DATASET_COLUMN_REFERENCE)) return plan
+
+    // We always remove the special metadata from `AttributeReference` at the end of this rule, so
+    // Dataset column reference only exists in the root node via Dataset transformations like
+    // `Dataset#select`.
+    val colRefs = plan.expressions.flatMap(_.collect {
+      case a: AttributeReference if isColumnReference(a) => toColumnReference(a)
+    })
+
+    if (colRefs.isEmpty) {
+      plan
+    } else {
+      // Keeps the mapping between the column reference and the actual column it points to. This
+      // will be used to replace the column references with actual columns in the root node later.
+      val colRefToActualCol = new mutable.HashMap[ColumnReference, AttributeReference]()
+      // Keeps the column references that points to more than one actual columns. We will not
+      // replace these ambiguous column references and leave them as they were.
+      val ambiguousColRefs = new mutable.HashSet[ColumnReference]()
```

Comment:
How can a column reference point to more than one actual columns? A column reference is determined by dataset id and column position. It can't have more than one column positions or dataset ids.

Reply:
e.g. `df1.join(df1).select(df1("id"))`

Reply:
It might point to multiple columns, but those columns are the same column. For example, `df1("id")` points to columns in both left and right plans, but the pointed columns are the same. Can a column reference point to different columns? If not, this can work without `ambiguousColRefs`, and the code can be simplified?

Reply:
think about outer joins. Columns from different join sides are always different columns.